src/modules/fw_pos_control_l1/FixedwingPositionControl.cpp

```diff
@@ -737,7 +737,23 @@ FixedwingPositionControl::set_control_mode_current(const hrt_abstime &now, bool
 
 	if (((_control_mode.flag_control_auto_enabled && _control_mode.flag_control_position_enabled) ||
 	     _control_mode.flag_control_offboard_enabled) && pos_sp_curr_valid) {
-		_control_mode_current = FW_POSCTRL_MODE_AUTO;
+		if (_pos_sp_triplet.current.type == position_setpoint_s::SETPOINT_TYPE_TAKEOFF) {
+			// TAKEOFF: handle like a regular POSITION setpoint if already flying
+			if (!in_takeoff_situation() && (_airspeed >= _param_fw_airspd_min.get() || !_airspeed_valid)) {
```

Comment:
>  this logic is copied twice now.. where before it was only there once, correct? Could we encapsulate? or maybe need to re-think the architecture a little.. seems that there are two partial state machines running in parallel now.

@tstastny Yes, This is exactly why I wanted to put the takeoff as a separate flight mode and deal with it there. I had to copy this because the takeoff waypoints are considered as normal "loiter setpoints" after 10 seconds.

IMHO, the fix is to make takeoff flight mode be it's own mode. But then I thought this should be done in a separate PR of itself. The state machines are not exactly running in parallel. There is only one state machine that decides whether to handle the takeoff waypoint as a position setpoint or a takeoff waypoint in `set_control_mode_current`